File: Native/tools/reference_original_audio_dsp.cpp
// The generated header wraps only the supplied reference numerical runStep,
// PACK andUNPACK bodies in an owned RAM object. No sound CPU, app startup,
// interrupts, emulator scheduler, JIT or host audio API is compiled or called.
#include "dsp_scalar_reference.generated.h"
#include "original_audio_dsp_bank.h"
#include "original_audio_dsp_test_input.h"
#include <filesystem>
#include <fstream>
#include <iostream>
#include <set>
#include <cstring>
using namespace idas3;
template<class T>void emit(std::ofstream& out,T value){out.write(reinterpret_cast<const char*>(&value),sizeof value);}
int main(int argc,char**argv){try{
    if(argc!=3)throw std::runtime_error("Supply project root and fixture output");
    std::filesystem::path root=argv[1];std::ofstream out(argv[2],std::ios::binary);out.write("ODSPRF01",8);
    struct Case {std::string name;unsigned preset;OriginalAudioDspProgram program;};
    std::vector<Case> cases;std::set<std::vector<std::uint8_t>> distinct;
    for(const auto& entry:std::filesystem::directory_iterator(root/"data/original_audio/dsp")){
        if(entry.path().extension()!=".idsp")continue;
        const auto name=entry.path().stem().string();const auto bank=loadOriginalAudioDspBank(root,name);
        for(unsigned i=0;i<bank.presets.size();++i){
            const auto& raw=bank.presets[i].originalRecord;std::vector<std::uint8_t> key(raw.begin(),raw.end());
            if(distinct.insert(key).second)cases.push_back({name,i,originalAudioDspProgram(bank,i,2,true)});
        }
    }
    emit(out,std::uint32_t(cases.size()));std::uint64_t codec=originalDspHashBasis;
    for(unsigned i=0;i<65536;++i)originalDspTestHash(codec,std::uint32_t(DspScalarReference::UNPACK(std::uint16_t(i))));
    for(unsigned i=0;i<0x1000000;++i)originalDspTestHash(codec,DspScalarReference::PACK(std::bit_cast<std::int32_t>(i<<8)>>8));
    emit(out,codec);unsigned count=0;
    for(const auto& [name,preset,program]:cases){
        constexpr unsigned frames=131072;DspScalarReference r(program),rotated(program);rotated.state.MDEC_CT=17329;
        std::uint64_t outputHash=originalDspHashBasis;std::uint32_t random=0x31415926;
        for(unsigned frame=0;frame<frames;++frame){
            const auto input=originalDspTestInput(random,frame);r.frame(input);rotated.frame(input);
            for(unsigned i=0;i<16;++i){
                if(std::int16_t(r.data.EFREG[i])!=std::int16_t(rotated.data.EFREG[i]))throw std::runtime_error("Original ring phase is observable");
                originalDspTestHash(outputHash,std::uint32_t(std::int32_t(std::int16_t(r.data.EFREG[i]))));
            }
        }
        std::uint64_t stateHash=originalDspHashBasis;
        for(auto v:r.state.TEMP)originalDspTestHash(stateHash,std::uint32_t(v));
        for(auto v:r.state.MEMS)originalDspTestHash(stateHash,std::uint32_t(v));
        for(unsigned i=0;i<65536;++i)originalDspTestHash(stateHash,unsigned(r.aica_ram[i*2])|(unsigned(r.aica_ram[i*2+1])<<8));
        originalDspTestHash(stateHash,r.state.MDEC_CT);
        char label[16]{};if(name.size()>=sizeof label)throw std::runtime_error("Fixture name extent");std::memcpy(label,name.data(),name.size());out.write(label,sizeof label);
        emit(out,std::uint32_t(preset));emit(out,std::uint32_t(frames));emit(out,outputHash);emit(out,stateHash);
        std::cout<<name<<": "<<frames<<" samples; output/state verified at two initial ring phases\n";++count;
    }
    if(!out)throw std::runtime_error("Cannot save scalar DSP fixture");std::cout<<"DSP reference: "<<count<<" distinct presets and all24-bit codec inputs.\n";
    return 0;
}catch(const std::exception& e){std::cerr<<e.what()<<'\n';return 1;}}

File: Native/tools/reference_original_audio_dsp.py
"""Build/run only local Flycast DSP arithmetic in an isolated owned-RAM shim.
The resulting fixture is consumed by portable native tests. No full runtime,
sound CPU, device mapping, instruction-cycle emulator or audio device is used.
"""
from pathlib import Path
import subprocess,hashlib,json
project=Path(__file__).resolve().parents[1];work=project.parents[1]/'work';work.mkdir(exist_ok=True)
reference=Path(r'C:/Users/Developer/Documents/Codex/reference_sources/flycast/core/hw/aica')
source=(reference/'dsp_interp.cpp').read_text();codec=(reference/'dsp.cpp').read_text()
def function(text,signature):
 start=text.index(signature);brace=text.index('{',start);depth=1;end=brace+1
 while depth:
  if text[end]=='{':depth+=1
  if text[end]=='}':depth-=1
  end+=1
 return text[start:end]
program=function(source,'void runStep()')
pack=function(codec,'u16 DYNACALL PACK').replace('u16 DYNACALL','static u16')
unpack=function(codec,'s32 DYNACALL UNPACK').replace('s32 DYNACALL','static s32')
header='''// Generated numerical-only reference. Original arithmetic source:
// Flycast dsp_interp.cpp (Audio Overload SDK, Copyright2007-2009 R. Belmont,
// Richard Bannister and others) and dsp.cpp (skmp/nullDC/reicast).
// Local source copyright/license remain applicable. Do not use as production code.
#pragma once
#include "original_audio_dsp.h"
#include <algorithm>
#include <cstring>
using u8=std::uint8_t;using u16=std::uint16_t;using u32=std::uint32_t;
using s16=std::int16_t;using s32=std::int32_t;using s64=std::int64_t;
struct DspScalarReference{
 struct State{s32 TEMP[128]{},MEMS[32]{},MIXS[16]{};u32 RBL=32767,RBP=0,MDEC_CT=1;bool stopped=false;}state;
 struct Data{u32 COEF[128]{},MADRS[64]{},MPRO[512]{},EXTS[2]{},EFREG[16]{};}data;
 Data*DSPData=&data;std::vector<u8>aica_ram=std::vector<u8>(131072);static constexpr u32 ARAM_MASK=131071;
 explicit DspScalarReference(const idas3::OriginalAudioDspProgram&p){
  state.RBL=p.ringLengthWords-1;
  for(unsigned i=0;i<128;++i){data.COEF[i]=u16(p.coefficients[i]);for(unsigned j=0;j<4;++j)data.MPRO[i*4+j]=p.instructions[i][j];}
  for(unsigned i=0;i<64;++i)data.MADRS[i]=p.memoryAddresses[i];
  for(unsigned i=0;i<65536;++i){u16 v=i<p.initialMemoryWords.size()?p.initialMemoryWords[i]:0x6000;aica_ram[i*2]=u8(v);aica_ram[i*2+1]=u8(v>>8);}
 }
 void frame(const std::array<s32,16>&input){std::copy(input.begin(),input.end(),state.MIXS);runStep();}
'''+pack+'\n'+unpack+'\n'+program+'\n};\n'
(work/'dsp_scalar_reference.generated.h').write_text(header)
out=project/'verification/original-audio-dsp';out.mkdir(parents=True,exist_ok=True)
cmd=work/'build_dsp_reference.cmd'
cmd.write_text('@echo off\ncall "C:\\Program Files\\Microsoft Visual Studio\\18\\Community\\VC\\Auxiliary\\Build\\vcvarsall.bat" x64 >nul\n'
 f'cl /nologo /std:c++20 /O2 /W4 /EHsc /I "{project / "src"}" /I "{project / "tests"}" /I "{work}" "{project / "tools/reference_original_audio_dsp.cpp"}" "{project / "src/original_audio_dsp_bank.cpp"}" /Fo:"{work}/" /Fe:"{work / "reference_original_audio_dsp.exe"}"\n'
 'if errorlevel 1 exit /b 1\n'
 f'"{work / "reference_original_audio_dsp.exe"}" "{project}" "{out / "scalar_reference.bin"}"\n')
subprocess.run(['cmd','/c',str(cmd)],check=True)
(out/'numerical_reference.json').write_text(json.dumps(dict(scope='Only supplied scalar DSP arithmetic; owned RAM, no hardware or full runtime. Used-program tests cover two-step read latency and odd-only SRAM operation ordering. Next-step write commit equals immediate write for this subset.',
 sources={n:hashlib.sha256((reference/n).read_bytes()).hexdigest() for n in ['dsp_interp.cpp','dsp.cpp']},
 notesSha256=hashlib.sha256((reference.parents[2]/'docs/neil_corlett_aica_notes.txt').read_bytes()).hexdigest(),
 fixtureSha256=hashlib.sha256((out/'scalar_reference.bin').read_bytes()).hexdigest(),
 samplesPerPreset=131072,initialRingPhases=[1,17329],codecUnpackInputs=65536,codecPackInputs=16777216,
excludedInstructions=['NOFL','TABLE','ADRL','ADREB','even-step SRAM operations','IWT without MRD two instructions before','same MEMS read/write bypass']),indent=2))

